rust/pact_ffi/src/verifier/mod.rs

```diff
@@ -147,6 +147,46 @@ ffi_fn! {
     }
 }
 
+ffi_fn! {
+    /// Set the verification options for the Pact verifier.
+    ///
+    /// `publish` is a boolean value. Set it to greater than zero to turn the option on.
+    /// `disable_ssl_verification` is a boolean value. Set it to greater than zero to turn the option on.
+    ///
+    /// # Safety
+    ///
+    /// All string fields must contain valid UTF-8. Invalid UTF-8
+    /// will be replaced with U+FFFD REPLACEMENT CHARACTER.
+    ///
+    fn pactffi_verifier_set_verification_options(
+      handle: *mut handle::VerifierHandle,
+      publish: c_uchar,
+      provider_version: *const c_char,
+      build_url: *const c_char,
+      disable_ssl_verification: c_uchar,
+      request_timeout: c_ulong,
+      provider_tags: *const *const c_char,
+      provider_tags_len: c_ushort
+    ) {
+      let handle = as_mut!(handle);
+      let provider_version = safe_str!(provider_version);
+      let build_url = safe_str!(build_url);
+
+      let tags = if !provider_tags.is_null() && provider_tags_len > 0 {
+        let mut tags = Vec::with_capacity(provider_tags_len as usize);
+        for index in 0..(provider_tags_len - 1) {
+          let tag_ptr: * const c_char = unsafe { *(provider_tags.offset(index as isize)) };
+          tags.push(safe_str!(tag_ptr).to_string());
+        }
+        tags
+      } else {
+        vec![]
+      };
```

Comment:
I can not extract this code into new function to avoid duplication. I am keep getting this error on build:

```
    |
569 |       tags.push(safe_str!(tag_ptr).to_string());
    |                 ^^^^^^^^^^^^^^^^^^ expected struct `Vec`, found enum `Result`
    |
```

Reply:
I'll need to see the code you extracted to understand this error

Reply:
This is the extracted function:
```rust
fn get_tags(provider_tags: *const *const c_char, provider_tags_len: c_ushort) -> Vec<String>
{
  let tags = if !provider_tags.is_null() && provider_tags_len > 0 {
    let mut tags = Vec::with_capacity(provider_tags_len as usize);
    for index in 0..(provider_tags_len - 1) {
      let tag_ptr: * const c_char = unsafe { *(provider_tags.offset(index as isize)) };
      tags.push(safe_str!(tag_ptr).to_string());
    }
    tags
  } else {
    vec![]
  };
  tags
}
```

Reply:
I don't know why that is happening, I get the same issue. The `safe_str` macro is doing something odd in that context.

Changing the function to the following works:
```rust
fn get_tags(provider_tags: *const *const c_char, provider_tags_len: c_ushort) -> Vec<String> {
  if !provider_tags.is_null() && provider_tags_len > 0 {
    let mut tags = Vec::with_capacity(provider_tags_len as usize);
    for index in 0..(provider_tags_len - 1) {
      let tag_ptr: *const c_char = unsafe { *(provider_tags.offset(index as isize)) };
      let tag = if_null(tag_ptr, "");
      if !tag.is_empty() {
        tags.push(tag.to_string());
      }
    }
    tags
  } else {
    vec![]
  }
}
```

Reply:
It works. Thanks